neqo-client/src/main.rs

```diff
@@ -150,30 +150,50 @@ fn process_loop(
             return client.state();
         }
 
-        let exiting = !handler.handle(args, client);
+        let mut exiting = !handler.handle(args, client);
 
-        let out_dgram = client.process_output(Instant::now());
-        emit_datagram(&socket, out_dgram.dgram());
+        loop {
+            let output = client.process_output(Instant::now());
+            match output {
+                Output::Datagram(dgram) => emit_datagram(&socket, Some(dgram)),
+                Output::Callback(duration) => {
+                    eprintln!("Timeout for {:?}", duration);
+                    socket.set_read_timeout(Some(duration)).unwrap();
+                    break;
+                }
+                Output::None => {
+                    eprintln!("Output::None");
+                    socket.set_read_timeout(None).unwrap();
+                    exiting = true;
+                    break;
+                }
+            }
+        }
 
         if exiting {
             return client.state();
         }
 
-        let sz = match socket.recv(&mut buf[..]) {
+        match socket.recv(&mut buf[..]) {
+            Err(ref err) if err.kind() == ErrorKind::WouldBlock => {
+                // timer expired
+                client.process_timer(Instant::now());
```

Comment:
If you take my suggestion from above, then you just need to do nothing on timer expiry.  The loop can return to process_output.